### ISSUE TITLE
Using "aziot_keys" as engine in the iothub_client_ll_x509_sample fails in connecting to IOT Hub (MQTT)

### ISSUE THREAD

**Issue by bemol38** [Azure/azure-iot-sdk-c#2268]
Hi everybody, I'm testing the use of the Azure IOT Identity Service together with the IotHub_client from IOT SDK C, in particular with the sample "iothub_client_ll_x509_sample". 
I have first to mention that the device provisioning to Azure done by the IOT Identity Service works so far well, obtaining the device certificate from an EST Server, and using TPM for key material storage.

The connection to the IOT Hub, using the engine "aziot_keys" in the sample software works sometimes, but not every time. I get the error: 
`../c-utility/adapters/tlsio_openssl.c Func:log_ERR_get_error Line:489 unable to load engine by ID`
when I start the sample software several times successive. How many times it works is unpredictable (1-2 sometimes 3 times)

To recover from this error, well, it seams that restarting the aziot service (sudo aziotctl system restart) helps, but it is not 100% sure.

I paid a lot of attention to always check the content of the key handle returned by the identity service, and which has to be put in the x509privatekey in the sample software, before running it again, because this key handle is changing often.
I can only assume, the OpenSSL engine "aziot_keys" for some reason is not well working together with the IOT SDK C ?
What could lead the engine to not be loaded by the SDK ?
I would happy to know about your opinion or experience regarding such an issue.

Thanks in advance.

**Comment by CIPop** [comment]
+@arsing since this is related to [Azure IoT Identity Service](https://github.com/Azure/iot-identity-service)'s OpenSSL Engine.

@bemol38 the C-SDK is calling into OpenSSL's engine APIs to load the aziot_keys engine. Are there any other error messages reported by C-SDK with details why this failed? (We query OpenSSL's `ERR_get_error()` that may return additional information.)

As far as I can tell from the description, it's very likely that the library is present in the correct location and load fails because the engine fails to initialize. The engine is connected to the Key Service (KeyD). 

On Ubuntu, the library is in `/usr/lib/x86_64-linux-gnu/engines-1.1/libaziot_key_openssl_engine_shared.so` and has a link present at `/usr/lib/x86_64-linux-gnu/engines-1.1/aziot_keys.so`.

@bemol38 I have two ideas you could try to get more information:

## Test if the issue is within C-SDK or the engine
When C-SDK fails to load the engine, please try using openssl s_client:

```sh
​openssl s_client -connect <endpoint> -cert ~/cert_from_identity_service.pem -keyform engine -engine aziot_keys -key "key_handle_from_identity_service"​
```
If the OpenSSL command fails, please close the issue here and re-open in https://github.com/Azure/iot-identity-service .

## Enable debug logging for KeyD

@arsing I haven't worked with Azure Identity Service for a while so please correct me if I'm wrong.

This is documented at https://github.com/Azure/iot-identity-service/blob/c4288e7ecc8ae978efa7a15ca71aebfec5c12cc3/docs-dev/running/aziot-keyd.md (step 5.)

```sh
export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$PWD/target/x86_64-unknown-linux-gnu/debug"
export AZIOT_LOG=aziot=debug

cargo run --target x86_64-unknown-linux-gnu -p aziotd -- aziot-keyd
```

Leave this running in another session and try the OpenSSL (or C-SDK) commands again. Please add the KeyD logs in a new issue in https://github.com/Azure/iot-identity-service.

**Comment by arsing** [comment]
>As far as I can tell from the description, it's very likely that the library is present in the correct location and load fails because the engine fails to initialize.

Yes, I agree.

>Enable debug logging for KeyD

(The docs-dev docs are for devs. Eg `cargo` is the build tool, users won't have it.)

Use `aziotctl system set-log-level debug` (or `iotedge system set-log-level` if you're using IoT Edge).

But in any case, yes, open an issue in the i-i-s repo instead (or you can transfer this one).

**Comment by bemol38** [comment]
Thank you everybody for your advices. I'll start investigate this beginning of Week 15.

**Comment by CIPop** [comment]
@bemol38 closing here. Please reopen in the Az IS repo.

**Comment by arsing** [comment]
( https://github.com/Azure/iot-identity-service/issues/393 )

**Comment by bemol38** [comment]
Hi @CIPop, could you please reopen this issue ?
The investigations according to your first recommendation ("**Test if the issue is within C-SDK or the engine**") on the Azure IoT Identity Service side are done and I closed the task [Azure/iot-identity-service#393](https://github.com/Azure/iot-identity-service/issues/393).
I need to investigate further.
Thank you in advance.

**Comment by bemol38** [comment]
Hi @CIPop,

Now openssl s_client is working. To make it work, I had to cleanup my system. There was several tpm2_pksc11 stores on my computer, some openssl conf files in many places. I also reinstall openssl, etc... 
Yesterday I recompiled azure iot csdk, and now the iotclient sample software modified to use aziot_keys engine is working perfectly.
So now, I think this issue can be closed.
